MIG-1045 Instructions for state-only transfer and CD migration (#45441)
openshift/openshift-docs: modules/migration-state-migration-cli.adoc

// * migrating_from_ocp_3_to_4/advanced-migration-options-3-4.adoc
// * migration_toolkit_for_containers/advanced-migration-options-mtc.adoc

:_content-type: PROCEDURE
[id="migration-state-migration-cli_{context}"]
= State migration

You can perform repeatable, state-only migrations by using {mtc-full} ({mtc-short}) to migrate persistent volume claims (PVCs) that constitute an application's state. You migrate specified PVCs by excluding other PVCs from the migration plan. You can map the PVCs to ensure that the source and the target PVCs are synchronized. Persistent volume (PV) data is copied to the target cluster. The PV references are not moved, and the application pods continue to run on the source cluster.

> State migration is specifically designed to be used in conjuction with external CD mechanisms, such as OpenShift Gitops. You can migrate application manifests using GitOps, while migrating the state using {mtc-short}.

Reviewer: ```suggestion
State migration is specifically designed to be used in conjunction with external CD mechanisms, such as OpenShift GitOps. You can migrate application manifests using GitOps while migrating the state using {mtc-short}.
```

Author: Done

> State migration is specifically designed to be used in conjuction with external CD mechanisms, such as OpenShift Gitops. You can migrate application manifests using GitOps, while migrating the state using {mtc-short}.

Reviewer 2: We try to stay away from the term "designed to" (https://www.ibm.com/docs/en/ibm-style?topic=word-usage#t). Maybe try "can"? or another variation. We also try to stay away from the term "by using". Let's try, for example: 

State migration can be used in conjunction with external CD mechanisms, such as OpenShift Gitops. You can migrate application manifests by using GitOps, while migrating the state by using {mtc-short}.

Author: @stevsmit 
'Can' does not really convey the idea. How about 'intended to be used'?

I am confused about 'by using'. There is no 'by using' in my sentence, but you introduced it in your suggestion.

Reviewer 2: Apologies @sbeskin-redhat . We try to stay away from the term "using", and suggest "by using" in its place. See https://www.ibm.com/docs/en/ibm-style?topic=word-usage#u for example. It's just a small nit. Sorry for the confusion.

> State migration is specifically designed to be used in conjuction with external CD mechanisms, such as OpenShift Gitops. You can migrate application manifests using GitOps, while migrating the state using {mtc-short}.

Reviewer 2: Also s/ conjuction / conjunction

> [id="migration-state-migration-cli_{context}"]

Author: Done

= State migration

You can perform repeatable, state-only migrations by using {mtc-full} ({mtc-short}) to migrate persistent volume claims (PVCs) that constitute an application's state. You migrate specified PVCs by excluding other PVCs from the migration plan. You can map the PVCs to ensure that the source and the target PVCs are synchronized. Persistent volume (PV) data is copied to the target cluster. The PV references are not moved, and the application pods continue to run on the source cluster.

> State migration is specifically designed to be used in conjuction with external CD mechanisms such as OpenShift Gitops. You can migrate application manifests using GitOps, while migrating the state using {mtc-short}.

Author: Done

====
State migration migrates only the components that constitute an application's state. If you want to migrate an entire namespace, use stage or cutover migration.
====

.Prerequisites

> * Application manifests are available in a central repository accessible by both the source and the target clusters.

Author: Done

* Application on the source cluster has persisted state in `PersistentVolumes` provisioned through `PersistentVolumeClaims`.

> .Recommended steps

Author: @mjpytlak 
I intentionally titled this section 'Recommended steps'. There are no real precise instructions here, just broad strokes. Do you still think it should be called 'Procedure'?

> .Recommended steps

Reviewer 3: > Do you think it should be called 'Procedure'?

Yeah. I understand where you are coming from, but I do believe we should call it `Procedure`.

. Migrate persistent volume data from the source to the target cluster.
+
You can perform this step as many times as needed. The source application continues running.

. Quiesce the source application.
+
You can do this by setting the replicas of workload resources to 0, either directly on the source cluster or by updating the manifests in GitHub before re-syncing the ArgoCD application.

. Clone application manifests to the target cluster.
+
You can use ArgoCD to clone the application manifests to the target cluster.

. Migrate the remaining volume data from the source to the target cluster.
+
Migrate any new data created by the application during steps 1 and 2 by performing a final data migration.

> . Un-quiesce the target application.

Author: Done

+
This step is only applicable if the cloned application is in quiesced state.

. Switch the DNS.
+

> Switch the DNS to the target cluster to re-direct user traffic to the migrated application.

Author: Done

====
State migration migrates only the components that constitute an application's state. If you want to migrate an entire namespace, use stage or cutover migration.
====

.Prerequisites

> * The application on the source cluster has its state persisted in `PersistentVolumes` provisioned through `PersistentVolumeClaims`.

Reviewer 4: You could update this prerequisite to use active voice. For example:

> * The state for the application on the source cluster is persisted in `PersistentVolumes` provisioned through `PersistentVolumeClaims`.

Author: @pneedle-rh  I've made this change but it's still passive voice.

> * The application's manifests are available in a central repository that is accessible by both the source and the target clusters.

Reviewer 4: You could remove the possessive from this line. For example:

> * The manifests for the application are available in a central repository that is accessible by both the source and target clusters.

.Procedure

. Migrate persistent volume data from the source to the target cluster.
+
You can perform this step as many times as needed. The source application continues running.

. Quiesce the source application.
+

> You can do this by setting the replicas of workload resources to 0, either directly on the source cluster or by updating the manifests in GitHub before re-syncing the ArgoCD application.

Reviewer 4: Because I'm not familiar with this topic area, I'm not sure from this line if the ArgoCD applications need to be re-synchronised only if the manifests are updated on GitHub, or if it is also needed if the replicas are set to 0 directly on the cluster. I'm assuming it's only required if the manifests are updated. Whichever it is, you could update this line to make it clear.

> You can do this by setting the replicas of workload resources to 0, either directly on the source cluster or by updating the manifests in GitHub and re-syncing the ArgoCD application.

Reviewer: ```suggestion
You can do this by setting the replicas of workload resources to `0`, either directly on the source cluster or by updating the manifests in GitHub and re-syncing the ArgoCD application.
```

> You can do this by setting the replicas of workload resources to 0, either directly on the source cluster or by updating the manifests in GitHub and re-syncing the ArgoCD application.

Reviewer 2: s/ 0/ `0`

Author: @stevsmit 
What does this mean?

Reviewer 2: s/ ArgoCD / Argo CD

Author: Done

. Clone application manifests to the target cluster.
+

> You can use ArgoCD to clone the application manifests to the target cluster.

Reviewer 2: Suggestion: 

The name of the application itself is "Argo CD". 

s/ ArgoCD / Argo CD

Author: Done

. Migrate the remaining volume data from the source to the target cluster.
+

> Migrate any new data created by the application during steps 1 and 2 by performing a final data migration.

Reviewer: Consider if it's necessary to mention the specific steps. There's not a way to automatically sync these references, so it's possible that they could become inaccurate after future updates.

```suggestion
Migrate any new data created by the application during the migration process by performing a final data migration.
```

> . If the cloned application is in quiesced state, unquiesce it.

Reviewer 4: You could update "quiesced state" to "a quiesced state".

Author: Done

> . If the cloned application is in a quiesced state, unquiesce it.

Reviewer 2: Suggestion: 

Unquiesce the cloned application if it is in a quiesced state.

> . Switch the DNS to the target cluster to re-direct user traffic to the migrated application.

Reviewer 4: Should "DNS" be "DNS entry" or something more specific than just "DNS"?

[NOTE]
====

> {mtc-short} 1.6 cannot quiesce applications automatically when performing state migration. It can only migrate PV data. Therefore, you must use your CD mechanisms for quiescing/unquiescing applications.

Reviewer 4: You could update "quiescing/unquiescing" to "quiescing or unquiescing".

Author: Done

> {mtc-short} 1.7 introduces explicit Stage and Cutover flows. You can use staging to perform initial data transfers as many times as needed. Then you can perform a cutover, in which the source applications are quiesced automatically.

Reviewer 4: In other areas of the MTC docs, I can see that "stage" and "cutover" are sometimes capitalised and sometimes not. You could consider making this consistent in the docs. This might be out of scope for this PR, but I thought I'd mention it anyway.

Author: I think Stage and Cutover are capitalized when they are flow names and not capitalized when they are actual actions.
